rfc/plugins.md

```diff
@@ -0,0 +1,343 @@
+
+In the case of a remote plugin, the plugin developer will have the freedom to choose language, framework, and runtime. 
+For any remote plugin, an in-process counterpart PF4J plugin will be required to enable communication between a service and the plugin. 
+For example, a remote plugin that exposes its integration through a gRPC server would also need a gRPC transport plugin on the service so that communication between the two can be configured correctly. 
+Other transport plugins, such as Docker co-process, Functions, HTTP, and so-on could also be developed.
+
+An extension point looks like the following:
+
+```kotlin
+// Definition of the PluginStage extension point.
+// The annotation defined here also informs the Spinnaker
+// plugin framwework which runtimes are acceptable. This
+// annotation would additionally be used for other metadata
+// Spinnaker needs to know atop whatever PF4J offers out of
+// the box.
+@SpinnakerExtension(allowRemote = true)
+class PipelineStageExtension : ExtensionPoint {
+  // ...
+}
+```
+
+It is highly recommended to read the documentation of PF4J to fully understand the scope of this proposal: This proposal will not go in depth on PF4J internals or taxonomy.
+
+### Front50 as the Source of Truth
+Operators will need the ability to understand, at a glance, what plugins are installed and enabled or disabled, their versions, as well as any other metadata about a plugin. 
+Whether plugins are implemented as in-process or not, Front50 should be used to track and expose a single source of truth of what plugins are installed in any particular service, as well as which ones are enabled or disabled.
+
+The Front50 plugin registry will also be exposed through Gate, and can be consumed by Deck to load the enabled plugins with UI components on application startup.
+
+Plugins can also be enabled or disabled at runtime without redeploying a service. 
+This flag is sourced from Front50, which all services will be listening for change events from. Whether this change event is published or polled for initially is unknown.
+
+Halyard configs can be used to inform Spinnaker of desired state, but whatever state is defined in Front50 will be used as the canonical source for plugins. 
+When a service starts up, it will ask Front50 what plugins need to be installed and which of those need to be enabled or disabled. Halyard’s role will only go as far as updating Front50’s repository of plugin metadata.
+
+Using Front50 as a source of truth will require extension of PF4J.
+
+```proto
+syntax = "proto3";
+
+// PluginManifest is used during registration of a plugin
+message PluginManifest {
+	// The name of the plugin, which is used as a fully 
+	// qualified identifier, including a namespace and 
+	// plugin name, ex: "netflix/fast-properties"
+	string name = 1;
```

Comment:
thoughts on making namespace be its own attribute? They are different attributes, so would be good to keep them separate. We've seen some issues when referencing plugin names via command line, see: https://spinnakerteam.slack.com/archives/CK9FK4XDF/p1571344728019300?thread_ts=1571286805.015200&cid=CK9FK4XDF
the issue above is that the slash is being URL encoded, which spring does not like to have in a request path. Modifying Halyard to send plugin names (including slashes) via query param works, but that does not seem like good API design

Reply:
Sure, that sounds fine to me. cc @jonsie

Reply:
cc @jonsie